src/compile/scale/parse.ts

```diff
@@ -80,20 +80,24 @@ function parseMainScale(model: Model, channel: Channel) {
   return scaleComponent;
 }
 
-function parseDomain(model: Model, fieldDef: FieldDef, channel: Channel): VgDomain {
+function parseDomain(model: Model, channel: Channel): VgDomain {
   const scale = model.scale(channel);
+
+  // FIXME(#): Can we always guarantee that the domain that we union is not already unioned?
+  type NonUnionDomain = any[] | VgDataRef;
+
   // If channel is either X or Y then union them with X2 & Y2 if they exist
   if (channel === X && model.channelHasField(X2)) {
     if (model.channelHasField(X)) {
       // FIXME: Verify if this is really correct
-      return { fields: [domain(scale, model, X), domain(scale, model, X2)] };
+      return { fields: [domain(scale, model, X) as NonUnionDomain, domain(scale, model, X2) as NonUnionDomain] };
```

Comment:
I have a much better solution. I extracted a unionDomains method that we can use here and in layer.ts.